File: HomeWork_3/src/com/valery/homework3/PartTwo/Action.java
package com.valery.homework3.PartTwo;

public abstract class Action {
    private Action() { }

    public static final class Login extends Action {
        private final User user;

        public Login(User user) {
            this.user = user;
        }

        public User getUser() {
            return user;
        }
    }

    public static final class Registration extends Action { }
    public static final class Logout extends Action { }
}

File: HomeWork_1/src/main/kotlin/Action.kt
sealed class Action

object Registration : Action()
class Login(val user: User) : Action()
object Logout : Action()

fun doAction(action: Action) = when(action) {
    is Login -> {
        println("Login has started\n")
        auth(action.user, ::updateCache)
    }
    is Logout -> println("Logout has started\n")
    is Registration -> println("Registration has started\n")
}

inline fun auth(user: User, completion: () -> Unit) = try {
    user.checkAge()
    authCallback.authSuccess()
    completion()
} catch (ex: Exception) {
    println(ex.message)
    authCallback.authFailed()
}

fun updateCache() {
    println("Cache has been updated\n")
}